TLS support

pjain1 opened the pull request:
Open for discussion.

Configurations -
1. `druid.server.http.plainText` - enable/disable http connector
2. `druid.server.http.tls`- enable/disable https connector
3. `druid.plainTextPort` - port for http connector
4. `druid.tlsPort` - port for https connector
5. `druid.server.https.*` - Jetty server TLS configs

Behavior of configuring http connector using `druid.host` and `druid.port` is kept backwards compatible. However, if `druid.server.http.tls` is set then `druid.tlsPort` config will be used for TLS port and if `druid.host` contains port then that port will be ignored, this also means that `druid.tlsPort` should be a non-negative Integer . Please see `DruidNodeTest` unit tests for all the different configurations that will work and ones that will not. 

High Level Implementation details -
1. Changed Internal Discovery metadata (`DruidServerMetadata`) to include `hostAndTlsPort` in addition to `host` which is combination of host and plain text port. If `druid.server.http.plainText` is not set then announced `host` will be null, similarly if `druid.server.http.tls`is not set then announced `hostAndTlsPort` will be null.
2. Changed Curator Announcer to announce `sslPort` in addition to `port`. `sslPort` will be -1 if tls is disabled, similarly `port` will be -1 if plainText is disabled. 
3. Changed `TaskLocation` to include `tlsPort`.
4. Changed `Worker` to include `scheme` which will be `https` if `tlsPort` is greater than or equal to 0 otherwise `http`.
5. Changed `ListenerResourceAnnouncer` to announce `HostAndPortWithScheme` instead of `HostAndPort`. If `tlsPort` is valid then host and tls port with `https` scheme will be used otherwise plainText port and `http` will be used.
6. Modified `PasswordProvider` to include `getPassword(key)` to support getting other passwords than just Metadata Storage Connector password. More details in comment in `PasswordProvider` class.
7. There are decent number of ways to configure SSLContext to be used with HttpClient. As per the code in master if an implementation of `SSLContext` is bound then that would be used with the client. Therefore, keeping it the same way but providing a sample implementation module (in extensions-core) that creates an example context, good enough for most simple cases.

TODO -
1. Tranquility support - May not be required immediately as added support for simultaneous HTTP and HTTPS connector.
2. Unit tests - Added few tests
~~3. Documentation~~
4. Possibly change integration tests to support TLS

I have tested the changes locally.

gianm commented:
@pjain1 very cool! Will fit in well with work we've been doing to add basic authentication, which require https to be effective…

How would a transition from http to https look for a cluster? For outwards-facing APIs, could they be served on both http and https for some transition period, before the user decides to cut over to https completely? I think something like that is necessary to avoid disrupting clients. (i.e. a three part migration: first enable both http/https on Druid, then switch clients to https, then turn off http on Druid)

For inwards-facing APIs, like historical, middleManager, task announcement, etc, what's the idea there? The fact that "scheme" is added but there's still only one "port" makes me think that they won't support both http and https at the same time. How would we do a rolling update in that case?

pjain1 commented:
We did not wanted to support two ports simultaneously on nodes to avoid complexity and code which will not be useful in future. This is what we thought about how rolling upgrade will work - 
1. Upgrade code on all nodes and put all other necessary infrastructure for certificates and keys. By default all service will keep on running using `http`.
2. Provide an implementation of `SSLContext` at all nodes (through an extension). Now the HttpClient on all nodes can talk to https enabled services if they are announcing scheme as https.
3. Now all the nodes can be upgraded one by one since nodes always look at the scheme to decide whether to talk on http or https.
4. External client have to handle https at some point so they should somehow handle this transition, probably they can have some logic where they can try connecting on https and if it doesn't work then fall back to http.

We thought of the three phase approach but thought it might a lot of unnecessary complexity to code which is not so nice.

pjain1 commented:
For the 4. point about external clients, actually they should also the external discovery to figure out the scheme and use that and then there won't be any disruptions.

gianm commented:
For the internally-facing APIs I think it's workable to only listen/announce one scheme at a time.

But for the externally-facing APIs I think users will really appreciate it if we listen on two ports for a while. That's because a _lot_ of peoples' clients don't use Curator service discovery to find Druid -- they use some other mechanism, like load balancers, VIPs, or even hard-coded URLs.

> External client have to handle https at some point so they should somehow handle this transition, probably they can have some logic where they can try connecting on https and if it doesn't work then fall back to http.

Some clients don't have the ability to use such custom logic, such as if your client is actually a load balancer or VIP where you can't configure a fallback like that. And even for clients that do have the ability to use custom logic, we'd be asking users to write code to implement that, which is a burden on them.

pjain1 commented:
Sounds like a valid concern. @himanshug @cheddar Do you have any opinions on Gian's comment.

himanshug commented:
@gianm it was discussed internally whether to enable same process to run both http and https simultaneously , but with scheme discovery via zookeeper it was possible to support druid's rolling upgrade and become https. so we decided not to go that route.
i think external clients can be taken care of if we add the feature on druid, so that if anyone sends `http://host:port/xxx` then server responds with "HTTP 301 Redirect" to `https://host:port/xxx` . Now, user would first upgrade the software one by one on all the external clients to handle https and the redirect... then enable https on druid side.... then change configuration on all the external clients to explicitly use https url.
does that sound ok ?

gianm commented:
Sorry but it still sounds too user-unfriendly to me.

Redirects aren't always easy or even possible for clients to follow. Consider the case where the brokers are fronted by a load balancer, which is super common (amazon ELB in particular is quite common, as are other kinds of load balancers). In many cases the user doesn't have access to the brokers directly and has to go through the balancer. If the load balancer doesn't follow redirects -- and many don't -- the client will get redirected to a url it can't access.

Also, many http libraries don't follow redirects by default, if they support it at all (our own http-client doesn't, and I've worked with others that also don't). And some libraries will follow for GET, but not for POST.

I think asking users to get all their clients to be able to deal with this kind of stuff before they can switch to https is going to be a pain for them, if it is even possible at all.

himanshug commented:
hmmm, in the case when brokers are behind a load balancer and all external clients talk to that ... then rolling upgrade is pretty simple ... remove `http://broker-1` from LB, then enable https on broker-1 and then add `https://broker-1` to LB ... and repeat same for all brokers.

that said, in the most general case only option would be to have broker accept both http and https. @cheddar any additional thoughts?

gianm commented:
That wouldn't work for Amazon ELBs, at least, which don't let you specify different schemes or ports for different instances behind a load balancer.

himanshug commented:
yeah, may be, so user would probably configure additional `druid.sslPort` which will be used to start additional jetty https connector.

gianm commented:
Yeah, that makes sense to me.

gianm commented:
We might want additional configs too to determine what ports to listen on. Something to make it possible to disable http when we don't want it anymore.

Would we also add 'standard' SSL ports for Druid services, meant to replace the 808x and 81xx ports we currently use?

pjain1 commented:
@gianm Ignoring the current configs in this PR, moving forward I was thinking of something like this -
1. If `tlsPort` is set and is different from `port` then two connectors will be opened one for http and other for https.
2. If `tlsPort` is set and is same as `port` or `port` is not set then use `tlsPort` to open https connector.
3. Otherwise use `port` to open http connector.
4. Fail in any other situation
In this world if users want https then they must set `tlsPort`.

Another design would be to have a config like `druid.server.https.enabled` -
1. Have same default value of `tlsPort` and ` port`
2. if the config is set and `tlsPort` and ` port` value is equal then open an https connector.
3. If the config is set and both ports values are different then open both http and https connector.
4. If the config is not set then use `port` value to open an http connector.

There can be multiple combinations with these 3 configs. What do you think is better ?

pjain1 commented:
One more option is to have a config like `druid.server.http.disable` (defaults to `false`) and `druid.server.https.enable` (defaults to `false`) . These two properties directly correlate to `port` and `tlsPort` on `DruidNode`. HTTP connector always uses `port` and HTTPS always uses `tlsPort`.

gianm commented:
Both of those seem a bit magical to me. If I'm a user I don't want to have to worry about whether `port` and `tlsPort` are the same and what implications that has.

Instead how about the following?

- set default tlsPort for each service to existing plaintext ports + 200; so broker is at 8282, historical is at 8283, MM tlsStartPort is at 8300, etc.
- have a config `druid.server.http.tls` which defaults to false, but if enabled, turns _on_ the tls ports.
- have a config `druid.server.http.plaintext` which defaults to false, but if disabled, turns _off_ the plaintext ports.
- users can customize port/tlsPort/startPort/tlsStartPort if they want, but we expect that most users won't.
- maybe deprecate the "port" configs and rename them to something like druid.plaintextPort?

gianm commented:
oops, should have said `druid.server.http.plaintext` would default to true. Edited the last comment.

pjain1 commented:
Interesting, my last comment looks similar to what you are proposing.

gianm commented:
Yes it does look similar :)

I like `druid.server.http.tls` and `druid.server.http.plaintext` -- to me it's easier to scan and understand something like `http.plaintext=true` than `http.disable=false`.

pjain1 commented:
OK lets go with the properties name you suggested. I'll make the changes soon. Thanks for the comments.

pjain1 commented:
If port is deprecated and if user was setting port using config then do we expect user to change it to plaintextPort or should the code internally map port to plaintextPort.

jon-wei commented:
@pjain1 @gianm 

Having two types of ports and a separate configuration to enable each type sounds good to me.

I think we could keep "port" as-is without deprecating, but if we do, I'd favor treating "port" and "plaintextPort" as equivalent.

If both plaintext/TLS ports are enabled, I think having `port` and `tlsPort` be the same should be an error condition instead of having TLS silently take precedence.

gianm commented:
> If port is deprecated and if user was setting port using config then do we expect user to change it to plaintextPort or should the code internally map port to plaintextPort.

In this case I'd expect the code to map port internally to plaintextPort. The idea would be that if a user isn't using tls, they shouldn't have to change anything in their config.

We don't necessarily need to deprecate port at all though, `druid.port` could basically mean "plaintext port". I'm not sure what's best, could go either way.

> If both plaintext/TLS ports are enabled, I think having port and tlsPort be the same should be an error condition instead of having TLS silently take precedence.

That does sound like it should be an error.

himanshug commented:
sounds like we are talking about 4 properties ...
```
druid.port or druid.plainTextPort
druid.tlsPort
druid.server.http.plaintext=true/false
druid.server.http.tls=true/false
```

how about having just 2 properties instead..
```
druid.port or druid.plainTextPort
druid.tlsPort
```

and have the behavior be following.
1. If nothing is specified, behavior stays same as now, only http connector is started on current default ports.
2. if either or both are specified then appropriate http connectors (http or https or both) are started.

gianm commented:
I think in the vast majority of cases we're really talking about just two properties, since most users will keep `druid.port` and `druid.tlsPort` at the default values.

One thing that seems weird to me about the "2 property" approach is that `druid.port` and `druid.tlsPort` have default values, and leaving the props out (i.e. keeping default values like 8282 for the broker) and explicitly specifying the same value (i.e. `druid.tlsPort=8282`) would have different behavior.

pjain1 commented:
I personally like 4 properties approach, makes things clear

himanshug commented:
@gianm with "2 properties" approach
- "default" case is when user does not say anything about ports in properties . and then http connector is started on current default port. note that no https connector is started in this case.
- if user specifies any port then there is no "default", system does exactly what is said in the config.  That is
  (1) if user only specified `druid.port` then only http connector is started.
  (2) if user only specified `druid.tlsPort` then only https connector is started.
  (3) if user specified both then both http and https connectors are started.

So, there are no default tls ports.

That said, "4 properties" might be a more commonly understood model as terradata has that too ( https://teradata.github.io/presto/docs/current/security/internal-communication.html ) and at least 2 people in this PR appear to like that more :) .

pjain1 commented:
@gianm Updated PR description as per the new design, please review. Fixed the code and added few unit tests as per the new design and is reviewable if you don't care about exhaustive unit tests.

nishantmonu51 commented on common/src/main/java/io/druid/metadata/PasswordProvider.java:
```diff
@@ -34,5 +34,14 @@
 public interface PasswordProvider
 {
+  // Used only for getting metadata storage connector password, present here for backwards compatibility
   public String getPassword();
```
deprecate this method so that it can be removed in future releases.

pjain1 replied:
done

nishantmonu51 commented on extensions-core/sample-client-sslcontext/README.md:
```diff
@@ -0,0 +1 @@
+This module is a sample implementation of SslContext that will be injected to be used with HttpClient.
```
can we also add some docs for SSLClientConfig and other configs needed to work with this module.

pjain1 replied:
ok

pjain1 replied:
Added documentation for all the server as well client configs

nishantmonu51 commented on indexing-service/src/main/java/io/druid/indexing/overlord/ForkingTaskRunner.java:
```diff
@@ -227,19 +232,25 @@ public TaskStatus call()
+                        }
+
+                        if(serverConfig.isTls()) {
+                          tlsChildPort = tlsPortFinder.findUnusedPort();
```
not a blocker for this, just to confirm the expected behavior. 
it seems separateIngestionEndpoint is not supported for TLS ? In that case, can we add that to docs

pjain1 replied:
Yes not supported with TLS as it is a deprecated config, I have added a note at line 404. Do you think it should be added here as well ?

nishantmonu51 replied:
thanks, no need to add it again here.

nishantmonu51 commented on server/src/main/java/io/druid/server/StatusResource.java:
```diff
@@ -55,7 +55,7 @@ public Status doGet()
     public Status(Collection<DruidModule> modules)
     {
-      this.version = getDruidVersion();
+      this.version = "0.9.2";
```
unrelated change ?

pjain1 replied:
reverted it back

jihoonson commented on server/src/main/java/io/druid/guice/MetadataConfigModule.java:
```diff
@@ -36,5 +42,36 @@ public void configure(Binder binder)
+            final PasswordProvider passwordProvider;
+            if (metadataStorageConnectorConfig == null || metadataStorageConnectorConfig.getPasswordProvider() == null) {
+              // this shouldn't happen as "druid.metadata.storage.connector.password" should always be set
+              // however it is useful for unit tests
```
Logging might be helpful to warn users that this should not happen.

pjain1 replied:
ok

jihoonson commented on server/src/main/java/io/druid/server/coordination/DruidServerMetadata.java:
```diff
@@ -27,7 +27,8 @@
   private final String name;
-  private final String host;
+  private final String hostAndPort;
+  private final String hostAndTlsPort;
```
Do we need to keep these concatenated strings? Maybe better to keep host, port, and tlsPort individually?

pjain1 replied:
This is used for announcing on ZK. So, wanted to keep announcement similar and just add one more field. I thought it might be easier to keep backward compatibility and less changes in code elsewhere. I don't know, if it is really required then I may take a shot at changing it.

nishantmonu51 commented on docs/content/development/extensions-core/sample-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+|`druid.client.https.algorithm`|SSL Algorithm/Protocol to use.|`null`|yes|
+|`druid.client.https.trustStoreType`|The type of the key store where trusted root certificates are stored.|`java.security.KeyStore.getDefaultType()`|yes|
+|`druid.client.https.trustStorePath`|The file path or URL of the TLS/SSL Key store where trusted root certificates are stored.|`null`|yes|
+|`druid.client.https.trustStoreAlgorithm`|Algorithm to be used by TrustManager to validate certificate chains|`javax.net.ssl.TrustManagerFactory.getDefaultAlgorithm()`|yes|
```
isn't this an optional setting with a default ? Required = No

pjain1 replied:
This is not optional. Why do you think so ?

nishantmonu51 replied:
since it seems it has a default value, so it is not mandatory for user to specify this value explicitly.

nishantmonu51 commented on distribution/pom.xml:
```diff
@@ -108,6 +108,8 @@
                                 <argument>-c</argument>
                                 <argument>io.druid.extensions:druid-examples</argument>
+                                <argument>-c</argument>
+                                <argument>io.druid.extensions:sample-client-sslcontext</argument>
```
is the sample-client-sslcontext intended to be used in production or users are encouraged to implement their own extensions ? 
If its intended for production use maybe drop the keyword 'sample' ?

pjain1 replied:
I guess it can be used for simple use cases. This extension can serve as an example for users for how to inject their own `SSLContext` if this extension does not satisfy their needs. 
@nishantmonu51 @gianm I need advice on this next point  -
Another motivation to implement it as an extension was to let the users understand what they are doing. For example users can point HTTPClient's `trustStore` to the same `keyStore` which server uses for storing its signed certificate and possibly some private keys. So, is it OK to do this ? I don't know so let the users decide and understand what they are doing. Does this reasoning make sense ?

pjain1 replied:
I guess TrustManager should take care of the isolation but not sure about the details of it.

jon-wei commented on docs/content/configuration/broker.md:
```diff
@@ -15,7 +15,8 @@ The broker node uses several of the global configs in [Configuration](../configu
 |--------|-----------|-------|
 |`druid.host`|The host for the current node. This is used to advertise the current processes location as reachable from another node and should generally be specified such that `http://${druid.host}/` could actually talk to this process|InetAddress.getLocalHost().getCanonicalHostName()|
-|`druid.port`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8082|
+|`druid.plainTextPort`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8082 if `druid.server.http.plainText` is set otherwise -1|
```
hm, I think we should keep `druid.port` so that if a user had a custom non-TLS port before this patch, that configuration wouldn't change

gianm replied:
`druid.port` definitely needs to keep working. Either `druid.plainTextPort` should be a synonym (and throw an error if user provides both and they aren't equal), _or_ `druid.port` should be the only one that exists.

pjain1 replied:
`druid.port` will still work, I just updated docs so that new users can start using `druid. plainTextPort` instead of `druid.port`

jon-wei commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+All these configs must be set to some valid value. Apart from these properties, `keyStorePassword` and `keyManagerPassword` are required
+to read Certificate from the Key Store. These passwords will be fetched using a `PasswordProvider` implementation.
+For example, `DefaultPasswordProvider` will look for `druid.password.keyStorePassword` and `druid.password.keyManagerPassword` runtime properties
+to get these passwords. `EnvironmentVariablePasswordProvider` will directly look for values of `keyStorePassword` and `keyManagerPassword` environment variables.
```
likewise here,  suggest "DRUID_KEYSTORE_PASSWORD" and "DRUID_KEY_MANAGER_PASSWORD" instead

jon-wei commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+Since, there are various ways to configure SSLContext, by default, Druid looks for an instance of SSLContext Guice binding
+while creating the HttpClient. This binding can be achieved writing a [Druid extension](../development/extensions.html)
+which can provide an instance of SSLContext. Druid comes with a sample extension present [here](../development/extensions-core/sample-client-sslcontext.html)
+which should be useful enough for most simple cases, see [this](./including-extensions.html) for how to include extensions.
```
since this extension is meant to be used in production for simple cases, I agree with @nishantmonu51 on removing "sample" (or maybe change it to "simple")

pjain1 replied:
changed

jon-wei commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java:
```diff
@@ -478,18 +484,34 @@ private FullResponseHolder submitRequest(
         throw new TaskNotRunnableException(String.format("Aborting request because task [%s] is not runnable", id));
       }
 
+      String host = location.getHost();
```
nit: i think you could make these final and set them later, since they aren't used until after their reassignment at line 497

pjain1 replied:
these are used in catch block so cannot be made final

jon-wei replied:
got it

gianm commented on docs/content/configuration/broker.md:
```diff
@@ -15,7 +15,8 @@ The broker node uses several of the global configs in [Configuration](../configu
 |`druid.host`|The host for the current node. This is used to advertise the current processes location as reachable from another node and should generally be specified such that `http://${druid.host}/` could actually talk to this process|InetAddress.getLocalHost().getCanonicalHostName()|
-|`druid.port`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8082|
+|`druid.plainTextPort`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8082 if `druid.server.http.plainText` is set otherwise -1|
+|`druid.tlsPort`|TLS port for HTTPS connector, if `druid.server.http.tls` is set then this config will be used. If `druid.host` contains port then that port will be ignored, this should be a non-negative Integer.|8282 if `druid.server.http.tls` is set otherwise -1|
```
A couple unanswered questions here:

- Where should `druid.server.http.tls` be set and what should it be set to? Should probably link to tls-support.html here.
- With the "8282 if `druid.server.http.tls` is set otherwise -1" bit, what does being set to -1 signify? That TLS is not being served?

gianm replied:
I think we don't really need the "otherwise -1" thing. The doc already says the `druid.tlsPort` is only going to be used if `druid.server.http.tls` is set.

pjain1 replied:
will link to tls-support page and remove "otherwise -1"

gianm replied:
thanks, that sounds good to me.

gianm commented on docs/content/development/extensions-core/sample-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+
+|Property|Description|Default|Required|
+|--------|-----------|-------|--------|
+|`druid.client.https.algorithm`|SSL Algorithm/Protocol to use.|`null`|yes|
```
What does `null` mean? What are some possible values for this?

gianm replied:
I think if you meant to say "there is no default" then "none" (without ticks) is more typical than `null`

pjain1 replied:
none would be better, thanks. I think I will update the code and doc to include a default algorithm.

pjain1 replied:
I can have "TLSv1.2" has the default. What do you think ?

gianm replied:
TLSv1.2 sounds like a good default.

gianm commented on docs/content/development/extensions-core/sample-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+|Property|Description|Default|Required|
+|--------|-----------|-------|--------|
+|`druid.client.https.algorithm`|SSL Algorithm/Protocol to use.|`null`|yes|
+|`druid.client.https.trustStoreType`|The type of the key store where trusted root certificates are stored.|`java.security.KeyStore.getDefaultType()`|yes|
```
Is this really required? If so what's the point of a default?

pjain1 replied:
yes this is required. Why do you think its not required ? 
For required fields, do you want me to not specify any defaults. I did it because `java.security.KeyStore.getDefaultType()` is something fetched on runtime so if users want to use this then not sure how to do that through a config.

jon-wei replied:
hm, if a default is offered, then the property shouldn't be required so users can omit the config

gianm replied:
Required means a user has to provide a value for the property (presumably, an exception will be thrown if the user doesn't provide one). Having a default means that the default will be used if the user doesn't provide a value. It's non-sensical for a field to be both required, and have a default, at the same time.

gianm replied:
It sounds like for this field, it makes sense for it to have a default, and not be required.

gianm commented on docs/content/development/extensions-core/sample-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+layout: doc_page
+---
+
+## Sample SSLContext Provider Module
```
This is a "sample". What situations do we expect users to use this in? Is it expected that they'll use it in production? Is it necessary to use an extension like this in order to get TLS to work?

pjain1 replied:
Yes it is necessary to use an extension for internal communication over TLS. I have mentioned his in the tls-support.md page under "Druid's internal communication over TLS" header.
For other comments - https://github.com/druid-io/druid/pull/4270#discussion_r120215855

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+|`druid.server.https.keyStoreType`|The type of the key store.|`null`|yes|
+|`druid.server.https.certAlias`|Alias of TLS/SSL certificate for the connector.|`null`|yes|
+
+All these configs must be set to some valid value. Apart from these properties, `keyStorePassword` and `keyManagerPassword` are required
```
I'm confused about the `keyStorePassword` and `keyManagerPassword` stuff. What's the default password provider? How do I change which password provider is used? Does changing the password provider affect other secrets I might have, like metadata store credentials, or S3 credentials, or anything related to kerberos?

gianm replied:
Also, what's the difference between a keyStorePassword and keyManagerPassword?

pjain1 replied:
`keyStorePassword ` is the password for the key store where the certificate resides whereas `keyManagerPassword` is the per key password, in this case it would be the password set for the certificate while storing it.

pjain1 replied:
`PasswordProvider` has been the interface used for getting metadata store password. `DefaultPasswordProvider` is the implementation used for fetching this password from runtime properties file. Users have to implement their own `PasswordProvider` in an extension id they want to has their own way of fetching the password.
Currently, this interface has just one method `getPassword()` which just assumes that it will be used to fetch metadata store password.
For the TLS support, Jetty server needs to access the key store which is password protected. I wanted a way to fetch this password which is extensible so that users can implement their own way for fetching pwd. So, I extended the `PasswordProvider` to include `getPassword(String key)` where key would be the thing using which the implementation decides which password to fetch. So, `keyStorePassword ` and `keyManagerPassword` are the keys. 
Probably, the confusion is `PasswordProvider` interface is not documented enough ? Do you have a better design in mind to achieve this functionality ?

gianm replied:
Ah, I see. Yeah, I was confused about how the PasswordProvider stuff was meant to work. I understand now but feel like it's not very intuitive; I wrote some comments on the interface just now.

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+
+# Druid's internal communication over TLS
+
+Whenever possible Druid nodes will use HTTPS to talk to each other. To enable this communication Druid's HttpClient need to
```
needs to (spelling)

pjain1 replied:
ok

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+
+Whenever possible Druid nodes will use HTTPS to talk to each other. To enable this communication Druid's HttpClient need to
+be configured with a proper [SSLContext](http://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html) that is able
+to validated the Server Certificates, otherwise communication will fail.
```
to validate (spelling)

pjain1 replied:
ok

pjain1 commented on extensions-core/lookups-cached-global/src/test/java/io/druid/query/lookup/NamespaceLookupExtractorFactoryTest.java:
```diff
@@ -442,7 +443,21 @@ public long getPollMs()
   public void testSerDe() throws Exception
   {
-    final Injector injector = makeInjector();
+    final Injector injector = Initialization.makeInjectorWithModules(
```
you are correct, makeInjector() already does this. Will change

pjain1 commented on extensions-core/lookups-cached-global/src/test/java/io/druid/query/lookup/NamespaceLookupExtractorFactoryTest.java:
```diff
@@ -472,7 +487,21 @@ public void testSerDe() throws Exception
   public void testSimpleIntrospectionHandler() throws Exception
   {
-    final Injector injector = makeInjector();
+    final Injector injector = Initialization.makeInjectorWithModules(
```
ok

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -0,0 +1,50 @@
+be configured with a proper [SSLContext](http://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLContext.html) that is able
+to validated the Server Certificates, otherwise communication will fail.
+
+Since, there are various ways to configure SSLContext, by default, Druid looks for an instance of SSLContext Guice binding
```
I think there's a common case where a user has their own in-house CA and uses that to sign server certificates. In this case, the client needs to know about the CA and use it. Is this case covered by the included sample extension? In this common case, should users use it in production?

If so, I think it shouldn't be called "sample" and should instead be called "simple". Just one letter changes, but now it sounds like something that could be used in production.

pjain1 replied:
I am not exactly sure of what you mean by in-house CA. Does this mean they have their own root certificates to validate the server's certificate chain ? If yes, then I think it is doable with this extension as users can point to a trustStore which has these root certificates and they specify the trustStoreType and algorithm used to validate certificate paths.

gianm replied:
Yes, I mean the user has their own root certificate(s) that should be used to verify the server certificate. I think it's worth calling out in the docs what this kind of user should do, since I think it will be really common.

pjain1 commented on extensions-core/sample-client-sslcontext/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    <dependency>
+      <groupId>com.metamx</groupId>
+      <artifactId>emitter</artifactId>
+      <version>0.4.1</version>
```
ok

jon-wei commented on server/src/main/java/io/druid/server/initialization/jetty/JettyServerModule.java:
```diff
@@ -169,20 +180,46 @@ static Server makeJettyServer(DruidNode node, ServerConfig config)
+      httpsConfiguration.setSecureScheme("https");
+      httpsConfiguration.setSecurePort(node.getTlsPort());
+      httpsConfiguration.addCustomizer(new SecureRequestCustomizer());
+      connectors[1] = new ServerConnector(
```
if `config.isPlainText()` is false but `config.isTls()` is true, connectors[] would have size 1 and this array access would be out of bounds

pjain1 replied:
fixed, thanks

jon-wei commented on docs/content/development/extensions-core/simple-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+
+|Property|Description|Default|Required|
+|--------|-----------|-------|--------|
+|`druid.client.https.algorithm`|SSL Algorithm/Protocol to use.|none|yes|
```
From your earlier comment about using "TLSv1.2" as a default, looks like this is referring to a protocol version under this section:
http://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#jssenames

Can you rename this to `druid.client.https.protocol` and remove "Algorithm" from the description? I found that somewhat confusing

pjain1 replied:
I named it algorithm because [this](http://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#SSLContext) mentions them as SSLContext algorithm and also the creation of SSLContext throws `NoSuchAlgorithmException`. However, I think I can change it to protocol

jon-wei replied:
Hm, the SSLContext documentation for getInstance calls the parameter a "protocol" but throws `NoSuchAlgorithmException` and points to the `SSLContext Algorithms` section you linked

In that case, I think either protocol/algorithm is fine, but it would be good if each property in this section had a direct link to the StandardNames section that corresponds with the property.

pjain1 commented on docs/content/development/extensions-core/sample-client-sslcontext.md:
```diff
@@ -0,0 +1,27 @@
+All these configs must be set to some valid value. Apart from these properties, `trustStorePassword` is required to access trustStore.
+This password will be fetched using a `PasswordProvider` implementation.
+For example, `DefaultPasswordProvider` will look for `druid.password.trustStorePassword` runtime property to get these passwords.
+`EnvironmentVariablePasswordProvider` will directly look for value of `trustStorePassword` environment variable.
```
`trustStorePassword` is the key name used to call `getPassword(String key)` method on `PasswordProvider` interface. So the key name sounds ok for `DefaultPasswordProvider` as it will actually look for `druid.password.trustStorePassword`. For `EnvironmentVariablePasswordProvider` we can have a convention which applies toUpper function to the key name and prefixes it with `DRUID_` to get the env name to look for, so it will look for "DRUID_TRUSTSTOREPASSWORD". What do you think ?

jon-wei replied:
Cool, the `DRUID_` + uppercasing convention for `EnvironmentVariablePasswordProvider` sounds good to me.

pjain1 commented on indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/PendingTaskBasedWorkerResourceManagementStrategy.java:
```diff
@@ -253,7 +253,7 @@ int getWorkersNeededToAssignTasks(
         // None of the existing worker can run this task, we need to provision one worker for it.
         // create a dummy worker and try to simulate assigning task to it.
-        workerRunningTask = createDummyWorker("dummy" + need, capacity, workerTaskRunnerConfig.getMinWorkerVersion());
+        workerRunningTask = createDummyWorker("http", "dummy" + need, capacity, workerTaskRunnerConfig.getMinWorkerVersion());
```
changed

pjain1 commented on indexing-service/src/main/java/io/druid/indexing/overlord/config/ForkingTaskRunnerConfig.java:
```diff
@@ -63,6 +63,11 @@
   private int startPort = 8100;
 
   @JsonProperty
+  @Min(1024)
```
ok

pjain1 commented on server/src/main/java/io/druid/server/DruidNode.java:
```diff
@@ -49,8 +48,26 @@
   @JsonProperty
-  @Min(0) @Max(0xffff)
-  private int port = -1;
+  @Max(0xffff)
```
this one cannot have @ min(1024) validation now as we allow -1 value for port in case user sets `plaintextPort` which is the new property. Probably I can put @ deprecated annotation here.

gianm replied:
`@Deprecated` with a comment about why sounds good to me.

pjain1 commented on server/src/main/java/io/druid/server/initialization/ServerConfig.java:
```diff
@@ -45,6 +45,12 @@
   private long maxScatterGatherBytes = Long.MAX_VALUE;
 
+  @JsonProperty
+  private boolean plainText = true;
```
ok

gianm commented on common/src/main/java/io/druid/metadata/PasswordProvider.java:
```diff
@@ -34,5 +34,15 @@
 public interface PasswordProvider
 {
+  // Used only for getting metadata storage connector password, present here for backwards compatibility
+  @Deprecated
```
This should be a javadoc comment.

gianm commented on common/src/main/java/io/druid/metadata/PasswordProvider.java:
```diff
@@ -34,5 +34,15 @@
 })
 public interface PasswordProvider
```
This interface changed from something metadata-store specific, into a global thing used to get all kinds of secrets. I think that's cool, but, how is the user supposed to set the global PasswordProvider they want to use? I see some Guice code binding it based on the value of `druid.metadata.storage.connector.password` but this seems kind of weird. Why would the global PasswordProvider be set through the metadata store properties? The ability to do so should be preserved for compatibility but there should also be a "nice" way to do it. Maybe `druid.secret.provider.type=whatever`.

Maybe consider this design instead.

- Add a new thing, SecretProvider, that is a global object for getting secrets.
- Make a SecretProviderPasswordProvider for the metadata store that causes the metadata connector to use the SecretProvider. Make this the default if no password is set for the metadata store.
- If someone sets `druid.metadata.storage.connector.password` then it would still create whatever kind of PasswordProvider they want, and that interface wouldn't change (but is really just there for legacy compatibility).

gianm replied:
Right now, it feels like you're wanting to make the PasswordProvider less specific to the metadata store (which I think is nice) but it's still too connected to really work well.

drcrallen replied:
As a note, we use the password provider interface to get secrets of various kinds, not just metadata. (See https://github.com/druid-io/druid/issues/3911 for similar)

drcrallen replied:
We also have internal implementations for `PasswordProvider` that do things like run blob storage object through cloud KMS decryption.

drcrallen replied:
Point being, changes to `PasswordProvider` should probably be in a PR of their own and not tied up in this one.

pjain1 replied:
@gianm What do you suggest the API for `SecretProvider` Interface look like, should this be same as PasswordProvider as per the changes in this PR, meaning are you just suggesting this design to decouple secret fetching logic from `MetadataStorageConnectorConfig` creation or something else ?
@drcrallen I'll do another PR for these changes

gianm replied:
I was thinking `SecretProvider` would look like the new version of `PasswordProvider` that you have in this PR. For this PR, I think it'd make sense to just do that and not touch `PasswordProvider` or metadata storage at all; then do a later PR to unify them.

pjain1 replied:
@cheddar also has some ideas on how to passwords should be fetched, lets discuss them in tomorrow's sync up and decide on something.

drcrallen commented on common/src/main/java/io/druid/metadata/PasswordProvider.java:
```diff
@@ -34,5 +34,15 @@
+   * {@param key} prefixed with {@link DefaultPasswordProvider#PWD_PROPERTY_PREFIX}
+   * in the runtime properties and return the value.
+   */
+  public String getPassword(String key);
```
I don't think this is a correct API for this, the configurations for the password provider can be supplied during the implementation construction, and having a "key" as a single object used to get passwords doesn't work in the general case. A simple example is an implementation that fetches an object from a blob storage, then uses a cloud KMS to decrypt that object. In this scenario you have both the blob storage path and the KMS coordinates that need configured, something that is not captured by a single `key` parameter.

jon-wei replied:
Hm, if key is regarded as a standardized "named secret" and not as the full configuration for accessing that secret, I think this interface could work.

Maybe the configuration for the custom PasswordProvider implementing the cloud KMS fetch+decrypt could have value mappings for "Druid-TLS-keyStorePassword" and ""Druid-TLS-keyManagerPassword" in a map somewhere, containing the blob storage paths and KMS coordinates for each "named secret"

gianm replied:
What @jon-wei said makes sense to me… I think the idea with this interface is that the `key` is a well-known config name like `metadataStorePassword` or `tlsKeyStorePassword` and it's up to the PasswordProvider to handle that however it wants to.

gianm replied:
Presumably we would have a table of these well-known key names somewhere.

jon-wei commented on server/src/main/java/io/druid/guice/StorageNodeModule.java:
```diff
@@ -65,6 +65,7 @@ public DruidServerMetadata getMetadata(@Self DruidNode node, @Nullable NodeTypeC
     return new DruidServerMetadata(
         node.getHostAndPort(),
         node.getHostAndPort(),
+        node.getHostAndTlsPort(),
```
For the first parameter here, `node.getHostAndPort()` can return null if plaintext ports are disabled but TLS ports are enabled, which can prevent nodes from announcing themselves to ZK (I saw this occur with a historical in a local test setup).

pjain1 replied:
Let me fix this and make sure same problem is not happening elsewhere where `getHostAndPort` is used.

jon-wei commented:
@pjain1 This patch mostly LGTM now, I'm okay with merging this once the discussions around the PasswordProvider are resolved.

I agree with @gianm and @drcrallen on keeping the existing password handling for the metadata store unchanged in this PR, and left my thoughts on @drcrallen's comment regarding the PasswordProvider(or SecretProvider) interface.

jon-wei commented on docs/content/configuration/indexing-service.md:
```diff
@@ -14,7 +14,8 @@ The indexing service uses several of the global configs in [Configuration](../co
 |--------|-----------|-------|
 |`druid.host`|The host for the current node. This is used to advertise the current processes location as reachable from another node and should generally be specified such that `http://${druid.host}/` could actually talk to this process|InetAddress.getLocalHost().getCanonicalHostName()|
-|`druid.port`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8090|
+|`druid.plaintextPort`|This is the port to actually listen on; unless port mapping is used, this will be the same port as is on `druid.host`|8090|
```
This should note that the MiddleManager defaults to ports 8091/8291

pjain1 replied:
added new section for MiddleManager configs

pjain1 commented:
@gianm @jon-wei @drcrallen to keep things simple I have kept `PasswordProvider` interface same and changed the code so that at each place where a password is required, a new object of `PasswordProvider` is instantiated. Thus each `PasswordProvider` can be individually set using runtime properties in similar way as it is done for the metadata store password. 

Added separate documentation pages for how to use and implement `PasswordProvider`, please have a look.

gianm commented:
@pjain1 the PasswordProvider approach looks good to me -- it's definitely simpler and looks like it should work well.

@jon-wei could you take another look at the overall patch please?

And @pjain1 could you please resolve the conflicts?

jon-wei commented:
@pjain1 Reviewed the patch again, didn't have any more comments, looks good to me. The new password provider design looks good.

I'll approve after resolving conflicts.

pjain1 commented:
I am going to do some basic testing again tomorrow after which I think the PR will be mergeable.

pjain1 commented:
The PR is mergeable now (if there are now more comments), note that I have added `toString` for `TLSServerConfig` and `SSLClientConfig` meaning that things like keyStorePath, keyStoreType etc. (excluding PasswordProviders) will be printed in logs.

jon-wei commented:
@drcrallen @nishantmonu51 Did you have any more comments on this patch?

jon-wei commented:
@nishantmonu51 @drcrallen @jihoonson @gianm 

Does anyone else want to take a look at this patch again? If not, I'm planning on merging this tomorrow.
